### PR TITLE
feat(table): 3715-Resize-Table-Column-header

### PR DESCRIPTION
**Pull request by sunilkotiyaibm**
#135

Closes #

Summary

Allow resizing Table columns using mouse dragging

Change List (commits, features, bugs, etc)

src/components/DataTable/_data-table.scss
src/components/Table/TableHead/TableHead.jsx
src/components/Table/TableHead/TableHeader.js
src/components/Table/TableHead/state/sorting.js
src/components/Table/TableHead/tools/cells.js
src/components/Table/TableHead/tools/sorting.js
src/components/Table/Table.test.jsx
src/components/Table/TableHead/TableHead.test.jsx
src/components/Dashboard/snapshots/Dashboard.story.storyshot
src/components/TableCard/snapshots/TableCard.story.storyshot
src/components/Table/snapshots/Table.story.storyshot
jest.config.js

Acceptance Test (how to verify the PR)

Allow resizing table column, on mouse hover resizing icon should be displayed
On resizing column using resizing icon should be dragged towards left or right direction using mouse down event
on releasing the mouse column should be resized where mouse was released and this should reflect in all rows.

### CODE REVIEW

**Comment by sunilkotiyaibm** [comment]
@davidicus please review this PR

File: src/components/Table/Table.test.jsx [hunk 1]
```diff
@@ -180,19 +180,19 @@ describe('Table', () => {
...
-        data={tableData}
-        actions={mockActions}
-        options={options}
-        view={view}
-      />
-    );
-    wrapper.find('button#column-string').simulate('click');
-    expect(mockActions.table.onChangeSort).toHaveBeenCalled();
-  });
+  // test('handles column sort', () => {
```
Comment (davidicus) [inline src/components/Table/Table.test.jsx]: @sunilkotiyaibm why are we commenting these  test out?

File: src/components/Table/TableHead/state/sorting.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+import { sortRows } from '../tools/sorting';
+
+/**
+ * We currently support the following sorting states for DataTable headers,
+ * namely: `NONE` for no sorting being applied, and then `DESC` and `ASC` for
+ * the corresponding direction of the sorting order.
+ */
+export const sortStates = {
```
Comment (davidicus) [inline src/components/Table/TableHead/state/sorting.js]: I don't think we need to bring this whole file in. Let's copy `sortStates` over in the `TableHeader.js` file that uses it.

File: src/components/Table/TableHead/tools/cells.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/**
```
Comment (davidicus) [inline src/components/Table/TableHead/tools/cells.js]: Did we mean to bring all of these extra files from Carbon?

File: src/components/Table/TableHead/tools/sorting.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/**
```
Comment (davidicus) [inline src/components/Table/TableHead/tools/sorting.js]: Do we need these files from Carbon?

File: src/components/Table/Table.test.jsx (hunk 1 above)
Comment (sunilkotiyaibm) [reply src/components/Table/Table.test.jsx]: This is uncommented now, after adding const sortStates = {
  NONE: 'NONE',
  DESC: 'DESC',
  ASC: 'ASC',
}; 
Now Table.test.jsx  test case is passing earlier it was failing

File: src/components/Table/TableHead/state/sorting.js (hunk 2 above)
Comment (sunilkotiyaibm) [reply src/components/Table/TableHead/state/sorting.js]: Done

File: src/components/Table/TableHead/tools/cells.js (hunk 3 above)
Comment (sunilkotiyaibm) [reply src/components/Table/TableHead/tools/cells.js]: Now removed, after using const sortStates = {
NONE: 'NONE',
DESC: 'DESC',
ASC: 'ASC',
};

File: src/components/Table/TableHead/tools/sorting.js (hunk 4 above)
Comment (sunilkotiyaibm) [reply src/components/Table/TableHead/tools/sorting.js]: Now removed, after using const sortStates = {
NONE: 'NONE',
DESC: 'DESC',
ASC: 'ASC',
};

**Comment by sunilkotiyaibm** [comment]
> 
> 
> I am still seeing the resize line off of its target in the preview link as well as getting into a condition where I could no longer resize.
> 
> 

Now issue is fixed, which is injected during RTL issue fix and wrongly commited.

**Comment by sunilkotiyaibm** [comment]
@davidicus Please review this PR #835

**Comment by davidicus** [comment]
Hi @sunilkotiyaibm. This is getting close. I am still seeing some inconsistencies in the preview link though. 

image

**Comment by sunilkotiyaibm** [comment]
> 
> 
> Hi @sunilkotiyaibm. This is getting close. I am still seeing some inconsistencies in the preview link though.
> 
> image
> 
> 

@davidicus , Ok we will re-look on this.

**Comment by sunilkotiyaibm** [comment]
> 
> 
> > Hi @sunilkotiyaibm. This is getting close. I am still seeing some inconsistencies in the preview link though.
> > image
> > 
> 
> @davidicus , Ok we will re-look on this.

@davidicus 2nd image issue fixed and delivered, 1st image issue not reproducible. 
Please review.

**Comment by kotiyas** [comment]
> Defect is still present
> 

@davidicus This behavior is due to Filter text, which is stopping resize process.
It will not be squeeze untill text box is exist. Here textbox have some fixed width because of that column will not be reseized, if you off the filter then will start resizing. Do we need to take care this filter issue in this story /feature ?
 As discussed wiith @JordanWSmith15 , we should disable the resizing when filter is on

**Comment by kotiyas** [comment]
> 
> Defect is still present
> 

@davidicus  we will look into this first image issue when height of header is increased and resize

**Comment by sunilkotiyaibm** [comment]
@JordanWSmith15  @davidicus we have fixed the resizing issue in first image, for filter scenario we need to finalize the behavior (here we can disable the filter if user resize, filter boxes have its own width so it resize the column when it is active again. ) Please review

**Comment by davidicus** [comment]
@sunilkotiyaibm I am not sure if that interaction works. Resizing these filter inputs seems like a prime use case for this feature. @JordanWSmith15  please feel free to weigh in. I am going to have Bjorn, whose been assigned to this issue, look into possible work arounds.

**Comment by JordanWSmith15** [comment]
> @sunilkotiyaibm I am not sure if that interaction works. Resizing these filter inputs seems like a prime use case for this feature. @JordanWSmith15 please feel free to weigh in. I am going to have Bjorn, whose been assigned to this issue, look into possible work arounds.

@davidicus  Agree that it is likely a use case, but given the length of time this story has gone on, I think it's time to draw the line and handle further capabilities as an add-on enhancement story, weighed against other priorities. I would be fine disabling resizing while the filters are active if it meant this story taking more than a few more days.

**Comment by stuckless** [comment]
@davidicus @sunilkotiyaibm @JordanWSmith15 Since we are talking able the filters... I also think that these filter components need to be customizable (like cell renderers and cell editors) since right now, if we have a date filter or a time filter, etc, we have no control over the user input component.  I'm also not a fan of the fact that the filters when enabled causes the table to grow quite considerably.  I think the expectation should be that the when a filter row is enabled the space that it gets is current column width.  I always find it disconcerting when I enable the filters and I feel like Iv'e lost columns, but then I realize it's because the table is now horizontal scrolling.

**Comment by davidicus** [comment]
> @davidicus @sunilkotiyaibm @JordanWSmith15 Since we are talking able the filters... I also think that these filter components need to be customizable (like cell renderers and cell editors) since right now, if we have a date filter or a time filter, etc, we have no control over the user input component. I'm also not a fan of the fact that the filters when enabled causes the table to grow quite considerably. I think the expectation should be that the when a filter row is enabled the space that it gets is current column width. I always find it disconcerting when I enable the filters and I feel like Iv'e lost columns, but then I realize it's because the table is now horizontal scrolling.

I agree. This is the behavior we are going with. Removing the min width and have these inputs be 100% of their cell size. Bjorn is pushing up some changes this morning with his fixes. 

@stuckless , custom sort filters are now a thing! Unfortunately, it seems to be a feature for Stateful table which i know you do not use. I wonder if that can be moved over to the base Table. We will have to look into. 
https://github.com/IBM/carbon-addons-iot-react/pull/843

File: src/components/Table/TableHead/TableHead.jsx [hunk 5]
```diff
@@ -128,6 +133,41 @@ const StyledCustomTableHeader = styled(TableHeader)`
...
+    minColumnWidth,
+    rightBound: colWidths[index],
+    leftBound: colWidths[index + 1] - minColumnWidth,
+  };
+};
+
+const updateColumnResizeWrapper = (validDrag, wrapperElement) => {
+  const invalidClass = 'column-resize-wrapper--invalid';
+  if (validDrag === false) {
+    wrapperElement.classList.add(invalidClass);
```
Comment (scottdickerson) [inline src/components/Table/TableHead/TableHead.jsx]: direct DOM manipulation like this seems like a bad idea, instead we should set an invalid state and Declaratively render the classname based on the state

File: src/components/Table/TableHead/TableHead.jsx [hunk 6]
```diff
@@ -148,9 +188,68 @@ const TableHead = ({
...
+  const mousemoveCallback = e => {
+    const mousePosition = e.clientX + columnVar.startX;
+    const direction = e.clientX > columnVar.move ? 'right' : 'left';
+    const dragBounds = getColumnDragBounds(direction, columnWidth, columnVar.index);
+    columnVar.validDrag =
+      document.dir === 'rtl'
+        ? dragIsValidRtl(mousePosition, dragBounds)
+        : dragIsValidLtr(mousePosition, dragBounds);
+    columnVar.element.style.left = `${mousePosition}px`;
+    updateColumnResizeWrapper(columnVar.validDrag, columnVar.element);
```
Comment (scottdickerson) [inline src/components/Table/TableHead/TableHead.jsx]: see above, I'd rather set a column state and have it declaratively choose the correct class accordingly

File: src/components/Table/TableHead/TableHeader.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  iconDescription: 'carbon.table.header.icon.description',
+};
+
+const translateWithId = (key, { sortDirection, isSortHeader }) => {
+  if (key === translationKeys.iconDescription) {
+    if (isSortHeader) {
+      // When transitioning, we know that the sequence of states is as follows:
+      // NONE -> ASC -> DESC -> NONE
+      if (sortDirection === sortStates.NONE) {
+        return 'Sort rows by this header in ascending order';
```
Comment (scottdickerson) [inline src/components/Table/TableHead/TableHeader.js]: this needs to be translated, right?

File: src/components/Table/TableHead/TableHead.jsx (hunk 5 above)
Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: I agree in general, but this class is added to the columnResizer element while it is being dragged and during that DOM action there is no react rendering going on.

File: src/components/Table/TableHead/TableHead.jsx (hunk 6 above)
Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: Me too, but also same reply as above :-)

File: src/components/Table/TableHead/TableHeader.js (hunk 7 above)
Comment (bjornalm) [reply src/components/Table/TableHead/TableHeader.js]: I don't know,  I think that @sunilkotiyaibm copied this file from carbon to temporarily add a forwarding ref.

Comment (davidicus) [reply src/components/Table/TableHead/TableHeader.js]: These ref changes need to be put into a PR on the Carbon repo still. I will open an issue to track.

File: src/components/Table/TableHead/TableHead.jsx (hunk 5 above)
Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: Sorry. Now I understand. I'll add the validDrag to the state to render while dragging. Will fix it first thing tomorrow.

Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: fixed

File: src/components/Table/TableHead/TableHead.jsx (hunk 6 above)
Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: done

File: src/components/Table/TableHead/TableHead.jsx [hunk 8]
```diff
@@ -148,9 +192,68 @@ const TableHead = ({
...
+    columnVar.element.style.left = null;
+    columnVar.validDrag = null;
+    updateColumnResizeWrapper(columnVar.validDrag, columnVar.element);
+  };
+  const onMouseDownCallback = (e, index) => {
+    columnVar.element = e.target;
+    columnVar.index = index;
+    columnVar.startX = columnVar.element.offsetLeft - e.clientX;
+    columnVar.move = e.clientX;
+    document.onmouseup = mouseupCallback;
```
[the remark this answers is not in the document]
Comment (bjornalm) [reply src/components/Table/TableHead/TableHead.jsx]: looking into it

File: src/components/Table/TableHead/TableHeader.js (hunk 7 above)
Comment (davidicus) [reply src/components/Table/TableHead/TableHeader.js]: > These ref changes need to be put into a PR on the Carbon repo still. I will open an issue to track.

[Issue](https://github.com/IBM/carbon-addons-iot-react/issues/862) created

**Comment by davidicus** [comment]
going to close in favor of #867

**Pull request closed without merging**